jedi/inference/gradual/annotation.py

```diff
@@ -331,27 +331,107 @@ def iter(iterable: Iterable[_T]) -> Iterator[_T]: ...
+        correctly cope with `Type[T]` annotations. When it is True, this means
+        that we are being called with a nested portion of an annotation and that
+        the `value_set` represents the types of the arguments, rather than their
+        actual instances.
+        Note: not all recursive calls will neccesarily set this to True.
     """
     type_var_dict = {}
+    annotation_name = annotation_value.py__name__()
+
+    def merge_pairwise_generics(annotation_value, annotated_argument_class):
+        """
+        Match up the generic parameters from the given argument class to the
+        target annotation.
+
+        This walks the generic parameters immediately within the annotation and
+        argument's type, in order to determine the concrete values of the
+        annotation's parameters for the current case.
+
+        For example, given the following code:
+
+            def values(mapping: Mapping[K, V]) -> List[V]: ...
+
+            for val in values({1: 'a'}):
+                val
+
+        Then this function should be given representations of `Mapping[K, V]`
+        and `Mapping[int, str]`, so that it can determine that `K` is `int and
+        `V` is `str`.
+
+        Note that it is responsibility of the caller to traverse the MRO of the
+        argument type as needed in order to find the type matching the
+        annotation (in this case finding `Mapping[int, str]` as a parent of
+        `Dict[int, str]`).
+
+        Parameters
+        ----------
+
+        `annotation_value`: represents the annotation to infer the concrete
+            parameter types of.
+
+        `annotated_argument_class`: represents the annotated class of the
+            argument being passed to the object annotated by `annotation_value`.
+        """
+
+        annotation_generics = annotation_value.get_generics()
+        actual_generics = annotated_argument_class.get_generics()
+
+        for annotation_generics_set, actual_generic_set in zip(annotation_generics, actual_generics):
+            for nested_annotation_value in annotation_generics_set:
+                _merge_type_var_dicts(
+                    type_var_dict,
+                    _infer_type_vars(
+                        nested_annotation_value,
+                        actual_generic_set,
+                        # This is a note to ourselves that we have already
+                        # converted the instance representation to its class.
+                        is_class_value=True,
+                    ),
+                )
+
     if isinstance(annotation_value, TypeVar):
         if not is_class_value:
-            return {annotation_value.py__name__(): value_set.py__class__()}
-        return {annotation_value.py__name__(): value_set}
+            return {annotation_name: value_set.py__class__()}
+        return {annotation_name: value_set}
     elif isinstance(annotation_value, TypingClassValueWithIndex):
```

Comment:
I think we could just make this an if (instead of elif) and use

```
annotation_generics = annotation_value.get_generics()
```

just before that, because it's used in every if branch AFAIK (albeit with different names).

Reply:
Unfortunately calling `annotation_value.get_generics()` at the top level breaks in the case where the `annotation_value` isn't generic. (Currently that branch drops out the bottom with `type_var_dict` untouched)

Do you mean having `annotation_generics = annotation_value.get_generics()` just inside the `if` block here? That would probably work, though I'm not sure it gains much in the current spelling of things.